src/services/outliningElementsCollector.ts

```diff
@@ -1,227 +1,166 @@
 /* @internal */
 namespace ts.OutliningElementsCollector {
-    const collapseText = "...";
-    const maxDepth = 40;
-    const defaultLabel = "#region";
-    const regionMatch = new RegExp("^\\s*//\\s*#(end)?region(?:\\s+(.*))?$");
-
     export function collectElements(sourceFile: SourceFile, cancellationToken: CancellationToken): OutliningSpan[] {
-        const elements: OutliningSpan[] = [];
-        let depth = 0;
-        const regions: OutliningSpan[] = [];
+        const res: OutliningSpan[] = [];
+        addNodeOutliningSpans(sourceFile, cancellationToken, res);
+        addRegionOutliningSpans(sourceFile, res);
+        return res.sort((span1, span2) => span1.textSpan.start - span2.textSpan.start);
+    }
 
-        walk(sourceFile);
-        gatherRegions();
-        return elements.sort((span1, span2) => span1.textSpan.start - span2.textSpan.start);
+    function addNodeOutliningSpans(sourceFile: SourceFile, cancellationToken: CancellationToken, out: Push<OutliningSpan>): void {
+        let depthRemaining = 40;
+        sourceFile.forEachChild(function walk(n) {
+            if (depthRemaining === 0) return;
+            cancellationToken.throwIfCancellationRequested();

```

Comment:
We are calling this on every single node -- we should come up with a way to call this less often.